parsing/parsetree.mli

```diff
@@ -71,90 +68,108 @@ and attributes = attribute list
 
-(* Type expressions *)
+(** {2 Type expressions} *)
 
 and core_type =
     {
      ptyp_desc: core_type_desc;
      ptyp_loc: Location.t;
      ptyp_loc_stack: location_stack;
-     ptyp_attributes: attributes; (* ... [@id1] [@id2] *)
+     ptyp_attributes: attributes;  (** [... [\@id1] [\@id2]] *)
     }
 
 and core_type_desc =
   | Ptyp_any
-        (*  _ *)
+        (** Represents [_] *)
   | Ptyp_var of string
-        (* 'a *)
+        (** Represents a type var such as ['a] *)
   | Ptyp_arrow of arg_label * core_type * core_type
-        (* T1 -> T2       Simple
-           ~l:T1 -> T2    Labelled
-           ?l:T1 -> T2    Optional
+        (** [Ptyp_arrow(lbl, T1, T2)] represents:
+            - [T1 -> T2]    when [lbl] is  {{!Asttypes.Nolabel}[Nolabel]},
+            - [~l:T1 -> T2] when [lbl] is {{!Asttypes.Labelled}[Labelled]},
+            - [?l:T1 -> T2] when [lbl] is {{!Asttypes.Optional}[Optional]}.
          *)
   | Ptyp_tuple of core_type list
-        (* T1 * ... * Tn
+        (** [Ptyp_tuple([T1 ; ... ; Tn])] represents a product type [T1 * ... * Tn].
 
-           Invariant: n >= 2
+           Invariant: [n >= 2].
         *)
   | Ptyp_constr of Longident.t loc * core_type list
-        (* tconstr
-           T tconstr
-           (T1, ..., Tn) tconstr
+        (** [Ptyp_constr(loc, l)] represents:
+            - [tconstr]               when [l=[]],
+            - [T tconstr]             when [l=[T]],
+            - [(T1, ..., Tn) tconstr] when [l=[T1 ; ... ; Tn]].
          *)
   | Ptyp_object of object_field list * closed_flag
-        (* < l1:T1; ...; ln:Tn >     (flag = Closed)
-           < l1:T1; ...; ln:Tn; .. > (flag = Open)
+        (** [Ptyp_object([ l1:T1; ...; ln:Tn ], flag)] Represents:
+            - [< l1:T1; ...; ln:Tn >]
+                                    when [flag] is {{!Asttypes.Closed}[Closed]},
+            - [< l1:T1; ...; ln:Tn; .. >]
+                                    when [flag] is   {{!Asttypes.Open}[Open]}.
          *)
   | Ptyp_class of Longident.t loc * core_type list
-        (* #tconstr
-           T #tconstr
-           (T1, ..., Tn) #tconstr
+        (** [Ptyp_class(tconstr, l)] represents:
+            - [#tconstr]               when [l=[]],
+            - [T #tconstr]             when [l=[T]],
+            - [(T1, ..., Tn) #tconstr] when [l=[T1 ; ... ; Tn]].
          *)
   | Ptyp_alias of core_type * string
-        (* T as 'a *)
+        (** Represents [T as 'a]. *)
   | Ptyp_variant of row_field list * closed_flag * label list option
-        (* [ `A|`B ]         (flag = Closed; labels = None)
-           [> `A|`B ]        (flag = Open;   labels = None)
-           [< `A|`B ]        (flag = Closed; labels = Some [])
-           [< `A|`B > `X `Y ](flag = Closed; labels = Some ["X";"Y"])
+        (** [Ptyp_variant([`A;`B], flag, labels)]Represents:
+            - [[ `A|`B ]]
+                 when [flag]   is {{!Asttypes.Closed}[Closed]}
+                  and [labels] is [None],
```

Comment:
I would use a pattern matching on the tuple directly:
```
when [flag], [labels] is [{{!Asttypes.Closed}[Closed]},  [None]
```

Reply:
Sorry, I wasn't clear enough. I meant to always use a pattern matching matrix, rather than resort to `and`, with the idea that this syntax is more readable than the plain text equivalent for OCaml programmers.

Reply:
Rereading the documentation on the generated html, the fully-phrased form  `when $x is ... and $y is ...` seems good enough. Since it is more important to converge at this stage, let's revert this pattern matching change and let's go with the `when ... and ...` form everywhere for this first version.

Reply:
Done